File: GO/nasa_bot.go
# go get github.com/go-telegram-bot-api/telegram-bot-api
# go run nasa_bot.go

package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math/rand"
	"net/http"
	"time"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api"
)

const (
	TELEGRAM_TOKEN = "YOUR_TELEGRAM_BOT_TOKEN"
	NASA_API_KEY   = "YOUR_NASA_API_KEY"
	NASA_API_URL   = "https://api.nasa.gov/planetary/apod?api_key=%s&date=%s"
)

func getRandomDate() string {
	currentYear := time.Now().Year()
	randomTime := time.Date(currentYear, time.Month(rand.Intn(12)+1), rand.Intn(28)+1, 0, 0, 0, 0, time.UTC)
	return randomTime.Format("2006-01-02")
}

func fetchNASAImage(date string) string {
	resp, err := http.Get(fmt.Sprintf(NASA_API_URL, NASA_API_KEY, date))
	if err != nil {
		return ""
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return ""
	}

	var result map[string]interface{}
	json.Unmarshal(body, &result)

	url, ok := result["url"].(string)
	if !ok {
		return ""
	}
	return url
}

func main() {
	bot, err := tgbotapi.NewBotAPI(TELEGRAM_TOKEN)
	if err != nil {
		panic(err)
	}

	u := tgbotapi.NewUpdate(0)
	u.Timeout = 60

	updates, err := bot.GetUpdatesChan(u)

	for update := range updates {
		if update.Message == nil {
			continue
		}

		date := getRandomDate()
		url := fetchNASAImage(date)

		if url == "" {
			msg := tgbotapi.NewMessage(update.Message.Chat.ID, "Sorry, couldn't fetch the photo.")
			bot.Send(msg)
			continue
		}

		photo := tgbotapi.NewPhotoUpload(update.Message.Chat.ID, url)
		bot.Send(photo)
	}
}

File: NASA_bot.lua
-- luarocks install lua-requests

local requests = require("requests")

local TELEGRAM_TOKEN = "YOUR_TELEGRAM_BOT_TOKEN"
local NASA_API_KEY = "YOUR_NASA_API_KEY"
local BASE_TELEGRAM_URL = "https://api.telegram.org/bot" .. TELEGRAM_TOKEN
local NASA_API_URL = "https://api.nasa.gov/planetary/apod?api_key=" .. NASA_API_KEY .. "&date="

local function getRandomDate()
    return os.date("%Y-%m-%d", math.random(os.time({year=os.date("%Y"), month=1, day=1}), os.time()))
end

local function fetchNASAImage(date)
    local response = requests.get(NASA_API_URL .. date)
    if response.status_code == 200 then
        local data = response.json()
        return data.url
    end
    return nil
end

local function getUpdates(offset)
    local response = requests.get(BASE_TELEGRAM_URL .. "/getUpdates?timeout=100&offset=" .. (offset or ""))
    if response.status_code == 200 then
        return response.json().result
    end
    return {}
end

local function sendMessage(chat_id, text)
    requests.get(BASE_TELEGRAM_URL .. "/sendMessage?chat_id=" .. chat_id .. "&text=" .. text)
end

local function sendPhoto(chat_id, photo_url)
    requests.get(BASE_TELEGRAM_URL .. "/sendPhoto?chat_id=" .. chat_id .. "&photo=" .. photo_url)
end

local offset

while true do
    local updates = getUpdates(offset)
    for _, update in ipairs(updates) do
        if update.message and update.message.text then
            local chat_id = update.message.chat.id
            local date = getRandomDate()
            local photo_url = fetchNASAImage(date)
            if photo_url then
                sendPhoto(chat_id, photo_url)
            else
                sendMessage(chat_id, "Sorry, couldn't fetch the photo.")
            end
        end
        offset = update.update_id + 1
    end
    os.execute("sleep " .. tonumber(1))
end
